Sources/AsyncHTTPClient/ConnectionPool.swift

```diff
@@ -0,0 +1,369 @@
+    /// by ensuring thread safety using the `lock`.
+    private subscript(key: Key) -> ConnectionProvider? {
+        get {
+            return self.lock.withLock {
+                _connectionProviders[key]
+            }
+        }
+        set {
+            self.lock.withLock {
+                _connectionProviders[key] = newValue
+            }
+        }
+    }
+
+    private enum RequiredAction {
+        case make(EventLoopPromise<ConnectionProvider>)
+        case existing(ConnectionProvider)
+    }
+
+    /// Determines what action should be taken regarding the the `ConnectionProvider`
+    ///
+    /// This method ensures there is no race condition if called concurently from multiple threads
+    private func getAction(for key: Key, eventLoop: EventLoop) -> RequiredAction {
+        return self.lock.withLock {
+            if let provider = self._connectionProviders[key] {
+                return .existing(provider)
+            } else {
+                let promise = eventLoop.makePromise(of: ConnectionProvider.self)
+                self._connectionProviders[key] = .future(promise.futureResult)
+                promise.futureResult.whenComplete { result in
+                    switch result {
+                    case .success(let provider):
+                        self[key] = provider
+                    case .failure:
+                        self[key] = nil
+                    }
+                }
+                return .make(promise)
+            }
+        }
+    }
+
+    /// This method asks the pool for a connection usable by the specified `request`, respecting the specified options.
+    ///
+    /// - parameter request: The request that needs a `Connection`
+    /// - parameter preference: The `EventLoopPreference` the connection pool will respect to lease a new connection
+    /// - parameter deadline: The connection timeout
+    /// - Returns: A connection  corresponding to the specified parameters
+    ///
+    /// When the pool is asked for a new connection, it creates a `Key` from the url associated to the `request`. This key
+    /// is used to determine if there already exists an associated `ConnectionProvider` in `connectionProviders`
+    /// if it does, the connection provider then takes care of leasing a new connection. If a connection provider doesn't exist, it is created.
+    func getConnection(for request: HTTPClient.Request, eventLoop: EventLoop, deadline: NIODeadline?) -> EventLoopFuture<Connection> {
+        let key = Key(request: request)
+
+        switch self.getAction(for: key, eventLoop: eventLoop) {
+        case .existing(let provider):
+            return provider.getConnection(eventLoop: eventLoop)
+
+        case .make(let providerPromise):
+            // If no connection provider exists for the given key, create a new one
+            let selectedEventLoop = providerPromise.futureResult.eventLoop
+            self[key] = .future(providerPromise.futureResult)
```

Comment:
Correct me if I'm missing something here, but we already write a promise to this key in `getAction` (on line 70), then we get the promise and set it again here in a racy fashion (without checking for a value for that key, just overwriting it)?

Reply:
Fixed it, thanks! I think the lock avoids any actual threading issue but it just didn't make sense and was redundant.

Reply:
Issue I think is not just about locks. `getConnection` itself is not under lock, so two threads with the same request could enter it, they both will get `.make` from `getAction`, they both will create a connection, but one of them will overwrite `self[key]`. On the other hand, even if we delete this write, we will overwrite `self[key`, but in the future complete closure

Reply:
@artemredkin / @adtrevor is this resolved?

Reply:
I still think this could be a race

Reply:
@artemredkin @weissi  I think https://github.com/swift-server/async-http-client/pull/105/commits/cae504ca07caaa9ae9fa0225f0b2bfc3965587d9 should fix this issue.